---
permalink: /
title: "Biography"
author_profile: true
redirect_from: 
  - /about/
  - /about.html
---


## What Can Be Found Here
Reason for these pages is to communicate results, thoughts and experience that not always end in papers that one can read afterwards. I can imagine it will be constantly under construction, but already now you can check the [Projects](https://mikekloccz.github.io/projects/) section where I give quick description about some recently finished or running projects. Some of the talks I have been giving are in section [Talks](https://mikekloccz.github.io/talks/). Some are really just technical about my experience with different bioinfomatics tools which I believe can be valuable to communicate in the community, but is exactly the case of things that will never get published by itself. If time allowes, I will add more "blog posts" about these topics in the future, and not just the pdf slides.


## About Me
*  *A data scientist working in computational cancer biology*. My job is to derive useful insights from "big data" that are generated in the lab or clinics to find new therapeutic or diagnostic options for patients. The tools I am using range from "the good old" statistical inference to modern machine learning approaches.

* Member  of [Momo Bentires-Alj](https://bentireslab.org/article-page/) research lab,   and also a member of *Swiss Institute of Bioinformatics* in the group of [Robert Ivanek](https://www.sib.swiss/directory/group/dbm-bioinformatics-core-facility).

*  One of the *founders of [RXcel](https://rxcel.ch)*, a pre-seed start-up where we aim at designing "smart" therapeutic RNAs to target aggressive solid tumors.

Next to the world of math, biology and data, there is another one that completely immerses me from time to time, and that's music. You can listen to some of my  production on [Soundcloud](https://soundcloud.com/michael-kloc).

## My Interests

There are roughly four main areas I am currently particularly interested in
*   What can we learn about [cell-cell interactions in single-cell](https://github.com/MikeKlocCZ/2025_scRNA_brain_mets_MassagueJ)  data how can we overlay multiple modalities, mainly with spatially-resolved data.

*  How can we infer reliably kinase activities from phosphoproteomic data to help identify novel therapeutic targets.

* How dynamical pattens in intracellular signaling determine cell fate or how precisely  we can model transcriptional regulation from sparse time-resolved sequencing or qPCR data.

* How "smart" RNA design allows to control optimal delivery profile of the effector protein.





<!-- *  Kinases are important players in intracellular signaling and can often be dysregulated (not only) in cancer. Inferring their activity from phosphoproteomic data is challenging, but a recently published [resource](https://bioconductor.org/packages/release/data/experiment/html/JohnsonKinaseData.html) for motif-based kinase scoring has significantly improved our ability to tackle this problem. Tested on data with known ground truth, dysregulated kinases were correctly identified, though false positives were common. To improve specificity, additional signaling patterns  and features "behind the linear motifs" should be integrated. 

* Recent advances in the field of optimal control have incorporated machine learning techniques into its standard toolkit. One concrete example is represented by neural ordinary differential equations (NODEs). This approach merges a mechanistic modelling of known dynamics in the form of ODEs with neural networks (NNs) embedded in it. Such models can be used to identify optimal control sequences that drive a system along a desired trajectory or to infer missing components of a mechanistic model from data. The approach can be extended to other types of differential equations (for example stochastic dynamics).  We are looking into these tools, for example, to help us understand dynamical interplay between transcription factors, or to decode complex patterns in cellular signaling.
 -->

## My Journey
I studied Physics at Charles University in Prague. My field of study was Nuclear and Subnuclear Physics. For my PhD I joint the group of [Prof. Pavel Cejnar](https://ipnp.cz/cejnar/publikace/clanky.html) where I specialized in Computational Quantum Physics of Many-Body Systems. After my PhD I was a visiting scientist at Technical University in Berlin and then I moved to Basel in 2019. I started as a post-doc in the group of [Prof. Christoph Bruder](https://quantumtheory-bruder.physik.unibas.ch/en/) where we worked on optimal control using tools of scientific machine learning. In 2021, I made a career shift towards biology and joined the lab of [Prof. Momo Bentires-Alj](https://bentireslab.org/article-page/). In 2023, I co-founded RXcel.
